### PR TITLE
✨ Add HostUpdatePolicy API

### PR DESCRIPTION
**Pull request by rhjanders**
**What this PR does / why we need it**:

This PR defines the API for setting HostUpdatePolicy. 

This is a part of Live Updates implementation as per design proposal: https://github.com/metal3-io/metal3-docs/pull/411

### CODE REVIEW

File: controllers/metal3.io/baremetalhost_controller.go [hunk 1]
```diff
@@ -1854,6 +1854,52 @@ func (r *BareMetalHostReconciler) createHostFirmwareSettings(info *reconcileInfo
...
+	if err := r.Get(info.ctx, info.request.NamespacedName, hup); err != nil {
+		if k8serrors.IsNotFound(err) {
+			// A resource doesn't exist, create one
+			hup.ObjectMeta = metav1.ObjectMeta{
+				Name:      info.host.Name,
+				Namespace: info.host.Namespace}
+
+			hup.Spec = metal3api.HostUpdatePolicySpec{
+				FirmwareSettings: metal3api.UpdatePolicy{}, 
+				FirmwareUpdates: metal3api.UpdatePolicy{}}
```
Comment (rhjanders) [inline controllers/metal3.io/baremetalhost_controller.go:1871]: This is clearly wrong - running tests results in:
```
janders@darkstar:~/git/hostupdatepolicy/baremetal-operator$ go test
# github.com/metal3-io/baremetal-operator/controllers/metal3.io
controllers/metal3.io/baremetalhost_controller.go:1870:23: invalid composite literal type "github.com/metal3-io/baremetal-operator/apis/metal3.io/v1alpha1".UpdatePolicy
controllers/metal3.io/baremetalhost_controller.go:1871:22: invalid composite literal type "github.com/metal3-io/baremetal-operator/apis/metal3.io/v1alpha1".UpdatePolicy
FAIL	github.com/metal3-io/baremetal-operator [build failed]
janders@darkstar:~/git/hostupdatepolicy/baremetal-operator$ 
```
likely missing something obvious - any pointers appreciated! :)

File: apis/metal3.io/v1alpha1/hostupdatepolicy_types.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	// defines policy for updating firmware
+	// +optional
+	// +kubebuilder:validation:Enum="onPreparing";"onReboot"
+	FirmwareUpdates UpdatePolicy `json:"firmwareUpdates,omitempty"`
+}
+
+// HostUpdatePolicyStatus defines the observed state of HostUpdatePolicy.
+type HostUpdatePolicyStatus struct {
+	// NOTE (janders) is this what we need to have empty status?
+	// if so I will remove the entire section
```
Comment (dtantsur) [inline apis/metal3.io/v1alpha1/hostupdatepolicy_types.go:47]: Yes, I think you can leave the empty structure.

File: controllers/metal3.io/baremetalhost_controller.go [hunk 3]
```diff
@@ -1855,6 +1855,23 @@ func (r *BareMetalHostReconciler) createHostFirmwareSettings(info *reconcileInfo
 	return nil
 }
 
+func (r *BareMetalHostReconciler) createHostUpdatePolicy(info *reconcileInfo) error {
+	// NOTE(janders) the goal here is to ensure that the controller reads the hup resource and adds OwnerReference to it
+	hup := &metal3api.HostUpdatePolicy{}
+	if !ownerReferenceExists(info.host, hup) {
```
Comment (dtantsur) [inline controllers/metal3.io/baremetalhost_controller.go:1861]: You're working with an empty object, you need to add loading (but not creating it).

File: apis/metal3.io/v1alpha1/hostupdatepolicy_types.go (hunk 2 above)
Comment (rhjanders) [reply apis/metal3.io/v1alpha1/hostupdatepolicy_types.go:47]: Thank you, removing the commented out code.

File: apis/metal3.io/v1alpha1/hostupdatepolicy_types.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+const (
+	HostUpdatePolicyOnPreparing    UpdatePolicy = "onPreparing"
+	HostUpdatePolicyOnReboot       UpdatePolicy = "onReboot"	
+)
+
+// NOTE(janders) based on https://github.com/metal3-io/metal3-docs/pull/411/
+
+// HostUpdatePolicySpec defines the desired state of HostUpdatePolicy
+type HostUpdatePolicySpec struct {
+	// defines policy for changing firmware settings
```
[the remark this answers is not in the document]
Comment (rhjanders) [reply apis/metal3.io/v1alpha1/hostupdatepolicy_types.go:33]: Done. Is there a linter that is commonly used among metal3 contributors that I should use to avoid mistakes like this in the future?

File: controllers/metal3.io/baremetalhost_controller.go (hunk 3 above)
Comment (rhjanders) [reply controllers/metal3.io/baremetalhost_controller.go:1861]: Fair point, now it feels I was a bit trigger happy deleting code from previous revision. Re-added the loading part, let's discuss if the code is good now.

File: apis/metal3.io/v1alpha1/hostupdatepolicy_types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (tuminoid) [reply apis/metal3.io/v1alpha1/hostupdatepolicy_types.go:33]: We have linters but they're not run on draft PR. As you can see there a lot of workflows as "skipped".

File: controllers/metal3.io/baremetalhost_controller.go [hunk 5]
```diff
@@ -1855,6 +1855,23 @@ func (r *BareMetalHostReconciler) createHostFirmwareSettings(info *reconcileInfo
 	return nil
 }
 
+func (r *BareMetalHostReconciler) createHostUpdatePolicy(info *reconcileInfo) error {
```
[the remark this answers is not in the document]
Comment (rhjanders) [reply controllers/metal3.io/baremetalhost_controller.go]: I spent some time trying to figure this out but haven't gotten far. createHostFirmwareSettings()/createHostFirmwareComponents() are called in registerHost() so that was my first thought but since we're not creating only setting owner probably not a good idea. Reconciler doesn't have reconcileInfo readily available. I'd appreciate a pointer or two.

[the remark this answers is not in the document]
Comment (hroyrh) [reply controllers/metal3.io/baremetalhost_controller.go]: Don't we need a reconciler for hostUpdatePolicy ?

[the remark this answers is not in the document]
Comment (rhjanders) [reply controllers/metal3.io/baremetalhost_controller.go]: OK I added a call from registerHost() to hostUpdatePolicySetOwnerReference().

File: apis/metal3.io/v1alpha1/hostupdatepolicy_types.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+// HostUpdatePolicy enumerates the allowed host update policies
+type UpdatePolicy string
+
+const (
+	HostUpdatePolicyOnPreparing    UpdatePolicy = "onPreparing"
+	HostUpdatePolicyOnReboot       UpdatePolicy = "onReboot"	
+)
+
+// NOTE(janders) based on https://github.com/metal3-io/metal3-docs/pull/411/
```
Comment (dtantsur) [inline apis/metal3.io/v1alpha1/hostupdatepolicy_types.go:29]: Please remove before finishing the draft

File: controllers/metal3.io/baremetalhost_controller.go [hunk 7]
```diff
@@ -151,7 +151,6 @@ func (r *BareMetalHostReconciler) Reconcile(ctx context.Context, request ctrl.Re
 	} else if hwdUpdated {
 		return ctrl.Result{Requeue: true}, nil
 	}
-
```
Comment (dtantsur) [inline controllers/metal3.io/baremetalhost_controller.go:154]: nit: unrelated whitespace change

File: controllers/metal3.io/baremetalhost_controller.go [hunk 8]
```diff
@@ -1855,6 +1860,27 @@ func (r *BareMetalHostReconciler) createHostFirmwareSettings(info *reconcileInfo
 	return nil
 }
 
+func (r *BareMetalHostReconciler) hostUpdatePolicySetOwnerReference(info *reconcileInfo) error {
+	// NOTE(janders) the goal here is to ensure that the controller reads the hup resource and adds OwnerReference to it
+	hup := &metal3api.HostUpdatePolicy{}
+	if err := r.Get(info.ctx, info.request.NamespacedName, hup); err != nil {
+		// Error reading the object
+		return errors.Wrap(err, "could not load hostUpdatePolicy resource")
```
Comment (dtantsur) [inline controllers/metal3.io/baremetalhost_controller.go:1868]: Please don't use the deprecated errors.Wrap, you need fmt.Errorf

File: apis/metal3.io/v1alpha1/hostupdatepolicy_types.go (hunk 6 above)
Comment (rhjanders) [reply apis/metal3.io/v1alpha1/hostupdatepolicy_types.go]: Done.

File: controllers/metal3.io/baremetalhost_controller.go (hunk 7 above)
Comment (rhjanders) [reply controllers/metal3.io/baremetalhost_controller.go]: Fixed.

File: controllers/metal3.io/baremetalhost_controller.go (hunk 8 above)
Comment (rhjanders) [reply controllers/metal3.io/baremetalhost_controller.go]: Done, thank you for the pointer.

**Comment by rhjanders** [comment]
I need to get to the bottom of the unit test failure.
/hold

File: controllers/metal3.io/baremetalhost_controller.go [hunk 9]
```diff
@@ -1855,6 +1859,27 @@ func (r *BareMetalHostReconciler) createHostFirmwareSettings(info *reconcileInfo
 	return nil
 }
 
+func (r *BareMetalHostReconciler) hostUpdatePolicySetOwnerReference(info *reconcileInfo) error {
+	// NOTE(janders) the goal here is to ensure that the controller reads the hup resource and adds OwnerReference to it
+	hup := &metal3api.HostUpdatePolicy{}
+	if err := r.Get(info.ctx, info.request.NamespacedName, hup); err != nil {
+		// Error reading the object
```
Comment (dtantsur) [inline controllers/metal3.io/baremetalhost_controller.go:1866]: Hmm, forgot, sorry: HTTP not found should be handled differently (i.e. ignored). See the other dependant resources for inspiration.

File: controllers/metal3.io/baremetalhost_controller.go [hunk 10]
```diff
@@ -1855,6 +1859,27 @@ func (r *BareMetalHostReconciler) createHostFirmwareSettings(info *reconcileInfo
...
+		return fmt.Errorf("could not load hostUpdatePolicy resource due to %w", err)
+	}
+	if !ownerReferenceExists(info.host, hup) {
+		if err := controllerutil.SetOwnerReference(info.host, hup, r.Scheme()); err != nil {
+			return errors.Wrap(err, "could not set bmh as owner for hostUpdatePolicy")
```
Comment (dtantsur) [inline controllers/metal3.io/baremetalhost_controller.go:1871]: nit: use fmt.Errorf here and below (Wrap is deprecated)

Comment (s3rj1k) [reply controllers/metal3.io/baremetalhost_controller.go:1871]: https://pkg.go.dev/errors#Join

File: controllers/metal3.io/baremetalhost_controller.go (hunk 9 above)
Comment (rhjanders) [reply controllers/metal3.io/baremetalhost_controller.go:1874]: Thank you Dmitry, very helpful pointer. I think I got it now, let's see what CI thinks (local tests passed).

**Comment by rhjanders** [comment]
Previously failing tests now pass
/unhold

File: controllers/metal3.io/baremetalhost_controller.go [hunk 11]
```diff
@@ -101,6 +101,9 @@ func (info *reconcileInfo) publishEvent(reason, message string) {
 // +kubebuilder:rbac:groups=metal3.io,resources=dataimages,verbs=get;list;watch;create;update;patch
 // +kubebuilder:rbac:groups=metal3.io,resources=dataimages/status,verbs=get;update;patch
 
+// Allow for updating hostupdatepolicies
+//+kubebuilder:rbac:groups=metal3.io,resources=hostupdatepolicies,verbs=get;list;watch;create;update;patch
```
Comment (dtantsur) [inline controllers/metal3.io/baremetalhost_controller.go:105]: I don't think BMO will ever edit these resources? I think having get,list,watch should be enough.

Comment (rhjanders) [reply controllers/metal3.io/baremetalhost_controller.go]: Good point. I adjusted this and force-pushed. Thank you!
